File: coolercontrold/src/sleep_listener.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2021-2025  Guy Boldon, Eren Simsek and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use crate::ENV_DBUS;
use anyhow::Result;
use log::{info, warn};
use moro_local::Scope;
use std::cell::Cell;
use std::env;
use std::ops::Not;
use std::rc::Rc;
use tokio_util::sync::CancellationToken;
use zbus::export::ordered_stream::OrderedStreamExt;
use zbus::{Connection, Proxy};

pub struct SleepListener {
    preparing_to_sleep: Rc<Cell<bool>>,
    resuming: Rc<Cell<bool>>,
}

impl<'s> SleepListener {
    pub async fn new(
        run_token: CancellationToken,
        scope: &'s Scope<'s, 's, Result<()>>,
    ) -> Result<Self> {
        let listener_enabled = env::var(ENV_DBUS)
            .ok()
            .and_then(|env_dbus| {
                env_dbus
                    .parse::<u8>()
                    .ok()
                    .map(|bb| bb != 0)
                    .or_else(|| Some(env_dbus.trim().to_lowercase() != "off"))
            })
            .unwrap_or(true);
        if listener_enabled.not() {
            info!("DBUS sleep listener disabled.");
            return Ok(Self::create_deaf_listener());
        }
        let conn_result = Connection::system().await;
        if conn_result.is_err() {
            // See issue:
            // https://gitlab.com/coolercontrol/coolercontrol/-/issues/264
            warn!("Could not connect to DBUS, sleep listener will not work!");
            return Ok(Self::create_deaf_listener());
        }
        let conn = conn_result?;
        let proxy = Proxy::new(
            &conn,
            "org.freedesktop.login1",
            "/org/freedesktop/login1",
            "org.freedesktop.login1.Manager",
        )
        .await?;

        let mut sleep_signal = proxy.receive_signal("PrepareForSleep").await?;
        let listener = Self {
            preparing_to_sleep: Rc::new(Cell::new(false)),
            resuming: Rc::new(Cell::new(false)),
        };
        let preparing_to_sleep = Rc::clone(&listener.preparing_to_sleep);
        let resuming = Rc::clone(&listener.resuming);
        scope.spawn(async move {
            loop {
                tokio::select! {
                    () = run_token.cancelled() => break,
                    Some(msg) = sleep_signal.next() => {
                        let body = msg.body();
                        let to_sleep: bool = body.deserialize()?; // returns true if entering sleep, false when waking
                        if to_sleep {
                            info!("Received message that system is going to sleep.");
                            preparing_to_sleep.set(true);
                        } else {
                            info!("Received message that system is waking from sleep");
                            resuming.set(true);
                        }
                    },
                    else => break,
                }
            }
            let _ = conn.close().await;
            Ok::<(), zbus::Error>(())
        });

        Ok(listener)
    }

    fn create_deaf_listener() -> Self {
        Self {
            preparing_to_sleep: Rc::new(Cell::new(false)),
            resuming: Rc::new(Cell::new(false)),
        }
    }

    pub fn is_resuming(&self) -> bool {
        self.resuming.get()
    }

    pub fn resuming(&self, is_resuming: bool) {
        self.resuming.set(is_resuming);
    }

    pub fn is_not_preparing_to_sleep(&self) -> bool {
        self.preparing_to_sleep.get().not()
    }

    pub fn preparing_to_sleep(&self, is_preparing_to_sleep: bool) {
        self.preparing_to_sleep.set(is_preparing_to_sleep);
    }
}

File: coolercontrol-gui/coolercontrol/services/sleep_listener.py
#  CoolerControl - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging.config
import time

from PySide6.QtCore import QThread
from apscheduler.job import Job
from jeepney import DBusAddress, MatchRule, message_bus
from jeepney.io.blocking import DBusConnection, open_dbus_connection, Proxy

from coolercontrol.settings import Settings, UserSettings

log = logging.getLogger(__name__)


class SleepListener(QThread):
    """
    This is a service that listens for a DBus message when the computer is put into sleep/hibernation and when it comes
    back out again. There are resets of some hardware systems when resuming and CoolerControl needs to re-apply
    all settings after waking up.
    This service extends and runs in its own QThread.
    """
    preparing_for_sleep_mode: bool = False
    _dbus_address_sleep: DBusAddress = DBusAddress('/org/freedesktop/login1',
                                                   bus_name='org.freedesktop.login1',
                                                   interface='org.freedesktop.login1.Manager')

    def __init__(self, device_update_jobs: list[Job]) -> None:
        super().__init__()
        self._device_update_jobs = device_update_jobs
        try:
            self._connection_system: DBusConnection = open_dbus_connection(bus='SYSTEM')
            self._match_rule = MatchRule(
                type="signal",
                interface=self._dbus_address_sleep.interface,
                member="PrepareForSleep",
                path=self._dbus_address_sleep.object_path
            )
            bus_proxy = Proxy(message_bus, self._connection_system)
            log.info("System DBus connection established: %s", bus_proxy.AddMatch(self._match_rule) == ())
            self.start()
        except BaseException as ex:
            log.error('Could not open DBus connection for listening', exc_info=ex)

    def run(self) -> None:
        with self._connection_system.filter(self._match_rule) as queue:  # if this errors out, terminate process
            while True:
                try:
                    log.debug("Listening...")
                    signal_msg = self._connection_system.recv_until_filtered(queue)
                    log.debug("DBus message received: %s ; %s", signal_msg.header, signal_msg.body)
                    if signal_msg.body[0]:  # returns true if entering sleep, false when waking
                        log.info("System is going to sleep/hibernating, pausing jobs")
                        SleepListener.preparing_for_sleep_mode = True
                        for job in self._device_update_jobs:
                            job.pause()
                    else:
                        log.info("System is resuming from sleep/hibernate, resuming jobs")
                        delay: int = Settings.user.value(UserSettings.STARTUP_DELAY, defaultValue=2, type=int)
                        # use startup delay in case usb connections take longer than normal
                        time.sleep(max(delay, 2))  # give the system at least a moment to wake up
                        log.debug("Resuming paused jobs after reinitialization")
                        for job in self._device_update_jobs:
                            job.resume()
                        time.sleep(1)  # give jobs a moment to process before waking up fully
                        SleepListener.preparing_for_sleep_mode = False
                except BaseException as ex:
                    log.error("Unexpected Error", exc_info=ex)

    def shutdown(self) -> None:
        if self.isRunning():
            self._connection_system.close()
            self.terminate()
            self.wait(3000)
        log.debug("Sleep DBus Listener shutdown")
